### PR TITLE
Transolver domain parallel

### PR DESCRIPTION
**Pull request by coreyjadams**
# PhysicsNeMo Pull Request

This PR implements adjustments to transolver to enable better, and more efficient, domain parallelism.  Batch size > 1 locally will not be supported for domain parallelism.

The key detail of these changes are to rearrange the data shapes.  Transolver takes inputs of shape `[Batch, Tokens, Features]`.  Previously, this was being converted to shapes of `[Batch, N_head, N_Tokens, Head_dim]`.  This pr essentialy replaces that with `[Batch, N_tokens, N_Head, Head_dim]` everywhere relevant.  The motivation for this is data contiguousness:  keeping the rearrange on the tensor dims with dim > sharded_dim is a 0-communication op, and everything stays nicely sharded.

Note as well, the attention mechanism is actually *not* parallel here (not that it impacts performance).  Check out the Transolver++ paper for details of their algorithm, but it's actually really easy to implement with ShardTensor - it just works.

https://arxiv.org/pdf/2502.02414v1

The idea is, the slice tokens are computed from a matmul of two pieces: `normed_weights`, and `fx`, and  the matmul contracts over the tokens dimension (which is the sharded dimension).  The output is a `Partial` placement, meaning each rank has a properly-shaped replica of the tokens but need to be all-reduced.

So, right before the attention layer, I implement a redistribute to a `Replicate()` placement which triggers the differentiable all reduce.  After the attention, the multiplication of `out_slice_tokens` and `slice_weights` produces a properly sharded outputs from the PhysicsAttention.  It's equivalent to this section of the upstream code (from the ++ model):
https://github.com/thuml/Transolver_plus/blob/main/models/Transolver_plus.py#L71-L73
but with one reduction, instead of two.

There is a boost to the attention path, too, in ShardTensor: it was computing ring Attention on replicated tensors which is slow (and wrong!) and now it is using a local attention on replicated tensors.

With these adjustments, I can efficiently scale the transolver model to about 3.6M input points on 8xA100.  

### Transolver ++

With these changes, we're actually only a hair's breadth away from having Transolver++.  I have the remaining updates on a separate branch where I want to validate training accuracy.

### Transformer Engine

Is not supported with domain parallelism.  The ops go through a different dispatch path - we need to implement that at some point, but it's not here.

### Tests

I added a new section of multi-gpu tests that are validating transolver end to end with shard tensor inputs.  It checks the outputs have the right shape, and the outputs are sharded properly.

### Changelog

I haven't updated it yet, I would like to bring Transolver ++ in, in it's entirety, and do it in one shot.

## Checklist

- [X] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

### CODE REVIEW

File: physicsnemo/models/transolver/Physics_Attention.py [hunk 1]
```diff
@@ -129,28 +133,48 @@ def compute_slices_from_projections(
             - The aggregated features are normalized by the sum of weights for numerical stability.
         """
```
Comment (pzharrington) [inline physicsnemo/models/transolver/Physics_Attention.py:139]: I think the tensor shapes in the docstring need to be updated to match the inline comments

File: physicsnemo/distributed/shard_utils/attention_patches.py [hunk 2]
```diff
@@ -649,7 +649,42 @@ def sdpa_wrapper(func: Callable, types: Any, args: tuple, kwargs: dict) -> Shard
...
+        raise MissingShardPatch("q, k, and v must all be on the same placement")
+
+    # Make sure the attention mask, if provided, has the same placement as q, k, and v
+    if attn_mask is not None and hasattr(attn_mask, "_spec"):
+        if attn_mask._spec.placements[0] != q._spec.placements[0]:
+            raise MissingShardPatch(
+                "attn_mask must have the same placement as q, k, and v"
+            )
+
+    # if the placements are replicatd (which is what we expect in transolver's Physics Attention)
```
Comment (pzharrington) [inline physicsnemo/distributed/shard_utils/attention_patches.py]: Small typo `replicatd`

File: physicsnemo/distributed/shard_utils/attention_patches.py [hunk 3]
```diff
@@ -649,7 +649,42 @@ def sdpa_wrapper(func: Callable, types: Any, args: tuple, kwargs: dict) -> Shard
...
+        local_output = torch.nn.functional.scaled_dot_product_attention(
+            local_q, local_k, local_v, attn_mask=local_attn_mask, **kwargs
+        )
+
+        output = ShardTensor.from_local(
+            local_output,
+            q._spec.mesh,
+            q._spec.placements,
+            # We don't have to worry about sharding shapes here since it's not sharded ...
+        )
```
Comment (pzharrington) [inline physicsnemo/distributed/shard_utils/attention_patches.py:685]: So this won't trigger CPU/GPU sync even though we left out the sharding shapes?

Comment (coreyjadams) [reply physicsnemo/distributed/shard_utils/attention_patches.py:685]: Correct, CPU/GPU sync is only triggered on sharded dimensions.
